Added 2 new features:  --batch and errorlevel 88

Creeper36 opened the pull request:
I added my comments to the Readme.md ....... I have been using this for 2 years now and I had to create 2 other python scripts to handle the log filter and another to SHA compare to find out if I got an update.  I decided to condense my 2 scripts into the original script (your great script).. and this is what I came up with... ALL other functions remain unchanged.. I did make a couple cosmetic word changes to reflect the automated batch script instead of the interactive mode or normal mode.

I feel others with batch scripts may find my edits useful and so thats why I have a pull request.. also.. I wouldnt have to manually update mine if you update the master script.

If you wanted to change my description or position in the Readme.md its understandable.. I put it at the top so you or others would see it first.

Thanks for your consideration!
Creeper

![batch](https://github.com/user-attachments/assets/1f20296d-5a5b-4b68-a45e-8c9583497ab3)

Creeper36 committed:
Added --batch, errorlevel 88, and updated README.md

Creeper36 committed:
layout / formatting fixed

OwenCochell commented:
Hello,

Apologies for the response delay, I have been traveling this past week. My response times should be greatly improved going forward.

Thank you for your contribution! This is an excellent idea and I would like to get it implemented. I think better features and documentation related to output and return codes is a good thing. I do have some comments on this implementation, and I will cover them here. Since you are using this script in your workflow, I would be curious to hear your thoughts on my proposed changes. I am keeping my feedback and thoughts pretty generic, but if you like I can markup your PR with comments in the relevant places.

# Return Code

Is there any reason for a return code of `88` after the script has installed the new server? I personally think it would be better to return a code of `0` when an operation has completed successfully. I would say 'success' in this case means that we did not encounter any unrecoverable errors, and all operations completed correctly. A return code of `0` usually means 'success' in most cases, so I think sticking to that would be good.

 I also think it would be useful if we return non-zero codes in certain scenarios, like if we preform a check and no new version is found. Just off the top of my head, there would be unique codes for the following situations:

- Network error
- Filesystem error
- File fails integrity check
- No new server version found

So in your case, if you get an error code of zero, then none of the above occurred, and you can continue knowing that a new version was found and installed without any problems. I think this makes more sense, offers more options for detecting various failures, and and aligns better with existing standards.

# Batch Implementation

I think this idea is very good! However, I worry that your proposed implementation might be bug-prone and hard to maintain. Keeping a list of patterns we don't care about is very unobtrusive (requires very little changes to the codebase), which is a good thing. However, I think this solution is not very explicit, and could lead to issues down the line if we change our style. Also, what if a user wanted to enable a single message type on their local copy that uses one of the patters, while filtering the others out? For example, if they wanted to see the message that states the script is working on version selection, but the user does not want to see the message related to version checking? Finally, your code will remove messages if they have filter strings at any position. While it is unlikely to occur for the filters you added, any message that has a filter string at any position will be removed. This is not very flexible and can be bug prone. 

Alternatively, I propose explicitly specifying what messages are 'batch related', and what messages are not required for this terse output mode. This is definitely more tedious, as changes would need to be made throughout the project to specify this for every message, but I think this approach does a better job as signaling what messages are important. It also allows for users to easily customize their copy of the software, as they can toggle importance on a per-message basis. 

There are a few ways to do this, an easy way is to alter the `output` function to take a boolean that determines if the message is terse enough to be shown. All calls to `output` will need to be changed to support this, but it's an option. Alternatively, we could rewrite the project to instead use the [python logging library](https://docs.python.org/3/library/logging.html). This would be a bigger feature to add but as of now I am leaning more towards this one.

# Conclusion

Sorry for the long message, I just wanted to hash out some ideas and comments. If you agree with me and would like to implement these changes on this PR, then we can figure that out. I'd be happy to markup your code and offer some possible implementations and feedback if you are interested. 

If you disagree with anything I said, then that's ok! We should talk about it and I would like to hear your input.

Otherwise, pending your feedback, I can implement these changes myself, using this comment as a starting point.

Thanks again!

Creeper36 commented:
Hello.. you really went into too much detail lol.. my whole reason for doing this was to share what I've already done.
RETURN CODES SECTION:   I used 88 cause it has personal meaning in batch files.. The main purpose of your python script is to update or not to update.. so you really only need to sys.exit on update or no update, 88 can be changed to whatever.. I dont want it as simple as 0 no update and 1 update.. thats too simple.. 0 and 8 would be okay.. since I use the batch and then redirect our scripts output to my logs it gives me a record of errors since they would still popup.. My only other possible errorlevel would be for '-u' ... but thats not necessary in batch mode cause it does say in the logs that the update did run and the status..   I did have 1 variation on this already completed:
================================================
if __name__ == '__main__':

 *********** CROPPED A BUNCH OUT *************

    if not args.check_only and update_available:

        # Allowed to install/Can install

        serv.get_new(default_version=args.version, default_build=args.build, backup=not (args.no_backup or args.new),
                    new=args.new, output_name=name, target_copy=args.copy_old)

        sys.exit(8)
        
    sys.exit(0)
^^^^^^^ This will exit with errorlevel 0 if no update is found
==============================================
BATCH SECTION: 

This was the best way I found to filter out what I didnt want to see in the logs.. its my personal creative choice.. its mainly the graphic, the names, the section titles, everything I left I decided was useful information for diagnostics.  I made sure the filters I added are very specific to certain outputs on the screen or in the logs.  All the filtering is for --batch users only.. no one will be looking at this unless theres a problem.
I wanted to keep my edits simple.. this is your creation I just wanted to share 2 examples of edits I actually implemented.. I really do like your script!  

I hope I answered some of your questions..  You can email any time.
Chad








    On Monday, January 13, 2025 at 06:05:37 PM EST, Owen Cochell ***@***.***> wrote:  
 
 


Hello,

Apologies for the response delay, I have been traveling this past week. My response times should be greatly improved going forward.

Thank you for your contribution! This is an excellent idea and I would like to get it implemented. I think better features and documentation related to output and return codes is a good thing. I do have some comments on this implementation, and I will cover them here. Since you are using this script in your workflow, I would be curious to hear your thoughts on my proposed changes. I am keeping my feedback and thoughts pretty generic, but if you like I can markup your PR with comments in the relevant places.

Return Code

Is there any reason for a return code of 88 after the script has installed the new server? I personally think it would be better to return a code of 0 when an operation has completed successfully. I would say 'success' in this case means that we did not encounter any unrecoverable errors, and all operations completed correctly. A return code of 0 usually means 'success' in most cases, so I think sticking to that would be good.

I also think it would be useful if we return non-zero codes in certain scenarios, like if we preform a check and no new version is found. Just off the top of my head, there would be unique codes for the following situations:
   
   - Network error
   - Filesystem error
   - File fails integrity check
   - No new server version found

So in your case, if you get an error code of zero, then none of the above occurred, and you can continue knowing that a new version was found and installed without any problems. I think this makes more sense, offers more options for detecting various failures, and and aligns better with existing standards.

Batch Implementation

I think this idea is very good! However, I worry that your proposed implementation might be bug-prone and hard to maintain. Keeping a list of patterns we don't care about is very unobtrusive (requires very little changes to the codebase), which is a good thing. However, I think this solution is not very explicit, and could lead to issues down the line if we change our style. Also, what if a user wanted to enable a single message type on their local copy that uses one of the patters, while filtering the others out? For example, if they wanted to see the message that states the script is working on version selection, but the user does not want to see the message related to version checking? Finally, your code will remove messages if they have filter strings at any position. While it is unlikely to occur for the filters you added, any message that has a filter string at any position will be removed. This is not very flexible and can be bug prone.

Alternatively, I propose explicitly specifying what messages are 'batch related', and what messages are not required for this terse output mode. This is definitely more tedious, as changes would need to be made throughout the project to specify this for every message, but I think this approach does a better job as signaling what messages are important. It also allows for users to easily customize their copy of the software, as they can toggle importance on a per-message basis.

There are a few ways to do this, an easy way is to alter the output function to take a boolean that determines if the message is terse enough to be shown. All calls to output will need to be changed to support this, but it's an option. Alternatively, we could rewrite the project to instead use the python logging library. This would be a bigger feature to add but as of now I am leaning more towards this one.

Conclusion

Sorry for the long message, I just wanted to hash out some ideas and comments. If you agree with me and would like to implement these changes on this PR, then we can figure that out. I'd be happy to markup your code and offer some possible implementations and feedback if you are interested.

If you disagree with anything I said, then that's ok! We should talk about it and I would like to hear your input.

Otherwise, pending your feedback, I can implement these changes myself, using this comment as a starting point.

Thanks again!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Creeper36 committed:
-- batch and Errorlevel 8 .. updated README.md

My Final Version For Pull Request

Creeper36 committed:
Some Final Edits on README.md

Creeper36 commented:
I've had a day to think about what you said and this is my followup #2 since I already sent you a 1st reply yesterday LOL.
-----------------------------        sys.exit(8)
                                                  << CHOICE 1
    sys.exit(0)-----------------------------        sys.exit(8)
    else:                                      << CHOICE 2        sys.exit(0)-----------------------------I personally like #2 cause its more clear and defined, but its a personal choice and can go either way correctly.
In your description you have errorlevel=0 being if no errors occured during update.. this thought process doesnt seem right to me.. I am not checking for errors to occur during this step.. i am checking if updates exist and to pass this along..  if no update =0.........if theres an update that completed successfully then exit with 8 to tell the batch file whats going on and pass along the results.. I have some post update scripting that I need this specific result for.  the whole purpose of this python script is to either update or no update.
as far as the --batch filtering results.. this is only visible during batch mode.. no other time will be supported since its whole purpose is to be used for batch scripts.. i want my logs small but still have all the info I need.. the current filter I setup is perfect for my needs.. in my opinion i filtered everything that a person using batch and capturing logs will need.
Again.. all this is already done and working 100%... I've tested it in just about every mode noting results/errors.. I took the time to study your script.. to really understand it.. and I really feel my changes are usefull for someone.. ME.... lol.. I hope others get use out of this but thats not why i started.  I consider your script complete as-is right now.. its very good.. no errors.. no hangs.. no crashes.. your kind of done creating it.. now its just time to maintain.. try not to overthink or make big changes.. it could be more unstable.. I've been using it unmonitored for 2 years.  implementing my edits to your script to get it to work how i want it to work.. I am just here to share and hope you implement my edits as-is but if you dont thats okay.. i will still maintain my fork.. i'd rather just use yours but i also need it to work for me.
if you fully implement my edits no one else will ever know.. it changes nothing for anyone currently using it.. its only there if you want to use it.
ThanksChad(your Indiana neighbor)


On Monday, January 13, 2025 at 06:05:37 PM EST, Owen Cochell ***@***.***> wrote: 




Hello,

Apologies for the response delay, I have been traveling this past week. My response times should be greatly improved going forward.

Thank you for your contribution! This is an excellent idea and I would like to get it implemented. I think better features and documentation related to output and return codes is a good thing. I do have some comments on this implementation, and I will cover them here. Since you are using this script in your workflow, I would be curious to hear your thoughts on my proposed changes. I am keeping my feedback and thoughts pretty generic, but if you like I can markup your PR with comments in the relevant places.

Return Code

Is there any reason for a return code of 88 after the script has installed the new server? I personally think it would be better to return a code of 0 when an operation has completed successfully. I would say 'success' in this case means that we did not encounter any unrecoverable errors, and all operations completed correctly. A return code of 0 usually means 'success' in most cases, so I think sticking to that would be good.

I also think it would be useful if we return non-zero codes in certain scenarios, like if we preform a check and no new version is found. Just off the top of my head, there would be unique codes for the following situations:
   
   - Network error
   - Filesystem error
   - File fails integrity check
   - No new server version found

So in your case, if you get an error code of zero, then none of the above occurred, and you can continue knowing that a new version was found and installed without any problems. I think this makes more sense, offers more options for detecting various failures, and and aligns better with existing standards.

Batch Implementation

I think this idea is very good! However, I worry that your proposed implementation might be bug-prone and hard to maintain. Keeping a list of patterns we don't care about is very unobtrusive (requires very little changes to the codebase), which is a good thing. However, I think this solution is not very explicit, and could lead to issues down the line if we change our style. Also, what if a user wanted to enable a single message type on their local copy that uses one of the patters, while filtering the others out? For example, if they wanted to see the message that states the script is working on version selection, but the user does not want to see the message related to version checking? Finally, your code will remove messages if they have filter strings at any position. While it is unlikely to occur for the filters you added, any message that has a filter string at any position will be removed. This is not very flexible and can be bug prone.

Alternatively, I propose explicitly specifying what messages are 'batch related', and what messages are not required for this terse output mode. This is definitely more tedious, as changes would need to be made throughout the project to specify this for every message, but I think this approach does a better job as signaling what messages are important. It also allows for users to easily customize their copy of the software, as they can toggle importance on a per-message basis.

There are a few ways to do this, an easy way is to alter the output function to take a boolean that determines if the message is terse enough to be shown. All calls to output will need to be changed to support this, but it's an option. Alternatively, we could rewrite the project to instead use the python logging library. This would be a bigger feature to add but as of now I am leaning more towards this one.

Conclusion

Sorry for the long message, I just wanted to hash out some ideas and comments. If you agree with me and would like to implement these changes on this PR, then we can figure that out. I'd be happy to markup your code and offer some possible implementations and feedback if you are interested.

If you disagree with anything I said, then that's ok! We should talk about it and I would like to hear your input.

Otherwise, pending your feedback, I can implement these changes myself, using this comment as a starting point.

Thanks again!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Creeper36 commented:
GitHub - Creeper36/PaperMC-Update at filter.cfg

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
GitHub - Creeper36/PaperMC-Update at filter.cfg

Contribute to Creeper36/PaperMC-Update development by creating an account on GitHub.
 |

 |

 |

 I made another branch for filter.cfg.


    On Monday, January 13, 2025 at 06:05:37 PM EST, Owen Cochell ***@***.***> wrote:  
 
 


Hello,

Apologies for the response delay, I have been traveling this past week. My response times should be greatly improved going forward.

Thank you for your contribution! This is an excellent idea and I would like to get it implemented. I think better features and documentation related to output and return codes is a good thing. I do have some comments on this implementation, and I will cover them here. Since you are using this script in your workflow, I would be curious to hear your thoughts on my proposed changes. I am keeping my feedback and thoughts pretty generic, but if you like I can markup your PR with comments in the relevant places.

Return Code

Is there any reason for a return code of 88 after the script has installed the new server? I personally think it would be better to return a code of 0 when an operation has completed successfully. I would say 'success' in this case means that we did not encounter any unrecoverable errors, and all operations completed correctly. A return code of 0 usually means 'success' in most cases, so I think sticking to that would be good.

I also think it would be useful if we return non-zero codes in certain scenarios, like if we preform a check and no new version is found. Just off the top of my head, there would be unique codes for the following situations:
   
   - Network error
   - Filesystem error
   - File fails integrity check
   - No new server version found

So in your case, if you get an error code of zero, then none of the above occurred, and you can continue knowing that a new version was found and installed without any problems. I think this makes more sense, offers more options for detecting various failures, and and aligns better with existing standards.

Batch Implementation

I think this idea is very good! However, I worry that your proposed implementation might be bug-prone and hard to maintain. Keeping a list of patterns we don't care about is very unobtrusive (requires very little changes to the codebase), which is a good thing. However, I think this solution is not very explicit, and could lead to issues down the line if we change our style. Also, what if a user wanted to enable a single message type on their local copy that uses one of the patters, while filtering the others out? For example, if they wanted to see the message that states the script is working on version selection, but the user does not want to see the message related to version checking? Finally, your code will remove messages if they have filter strings at any position. While it is unlikely to occur for the filters you added, any message that has a filter string at any position will be removed. This is not very flexible and can be bug prone.

Alternatively, I propose explicitly specifying what messages are 'batch related', and what messages are not required for this terse output mode. This is definitely more tedious, as changes would need to be made throughout the project to specify this for every message, but I think this approach does a better job as signaling what messages are important. It also allows for users to easily customize their copy of the software, as they can toggle importance on a per-message basis.

There are a few ways to do this, an easy way is to alter the output function to take a boolean that determines if the message is terse enough to be shown. All calls to output will need to be changed to support this, but it's an option. Alternatively, we could rewrite the project to instead use the python logging library. This would be a bigger feature to add but as of now I am leaning more towards this one.

Conclusion

Sorry for the long message, I just wanted to hash out some ideas and comments. If you agree with me and would like to implement these changes on this PR, then we can figure that out. I'd be happy to markup your code and offer some possible implementations and feedback if you are interested.

If you disagree with anything I said, then that's ok! We should talk about it and I would like to hear your input.

Otherwise, pending your feedback, I can implement these changes myself, using this comment as a starting point.

Thanks again!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Creeper36 commented:
sorry for all the replies like this.. this is how i think... as I get an idea i have to send it out LOL.
i do not like the filter.cfg idea but I wanted to give you options.. I would still prefer it built into the .py script.









    On Monday, January 13, 2025 at 06:05:37 PM EST, Owen Cochell ***@***.***> wrote:  
 
 


Hello,

Apologies for the response delay, I have been traveling this past week. My response times should be greatly improved going forward.

Thank you for your contribution! This is an excellent idea and I would like to get it implemented. I think better features and documentation related to output and return codes is a good thing. I do have some comments on this implementation, and I will cover them here. Since you are using this script in your workflow, I would be curious to hear your thoughts on my proposed changes. I am keeping my feedback and thoughts pretty generic, but if you like I can markup your PR with comments in the relevant places.

Return Code

Is there any reason for a return code of 88 after the script has installed the new server? I personally think it would be better to return a code of 0 when an operation has completed successfully. I would say 'success' in this case means that we did not encounter any unrecoverable errors, and all operations completed correctly. A return code of 0 usually means 'success' in most cases, so I think sticking to that would be good.

I also think it would be useful if we return non-zero codes in certain scenarios, like if we preform a check and no new version is found. Just off the top of my head, there would be unique codes for the following situations:
   
   - Network error
   - Filesystem error
   - File fails integrity check
   - No new server version found

So in your case, if you get an error code of zero, then none of the above occurred, and you can continue knowing that a new version was found and installed without any problems. I think this makes more sense, offers more options for detecting various failures, and and aligns better with existing standards.

Batch Implementation

I think this idea is very good! However, I worry that your proposed implementation might be bug-prone and hard to maintain. Keeping a list of patterns we don't care about is very unobtrusive (requires very little changes to the codebase), which is a good thing. However, I think this solution is not very explicit, and could lead to issues down the line if we change our style. Also, what if a user wanted to enable a single message type on their local copy that uses one of the patters, while filtering the others out? For example, if they wanted to see the message that states the script is working on version selection, but the user does not want to see the message related to version checking? Finally, your code will remove messages if they have filter strings at any position. While it is unlikely to occur for the filters you added, any message that has a filter string at any position will be removed. This is not very flexible and can be bug prone.

Alternatively, I propose explicitly specifying what messages are 'batch related', and what messages are not required for this terse output mode. This is definitely more tedious, as changes would need to be made throughout the project to specify this for every message, but I think this approach does a better job as signaling what messages are important. It also allows for users to easily customize their copy of the software, as they can toggle importance on a per-message basis.

There are a few ways to do this, an easy way is to alter the output function to take a boolean that determines if the message is terse enough to be shown. All calls to output will need to be changed to support this, but it's an option. Alternatively, we could rewrite the project to instead use the python logging library. This would be a bigger feature to add but as of now I am leaning more towards this one.

Conclusion

Sorry for the long message, I just wanted to hash out some ideas and comments. If you agree with me and would like to implement these changes on this PR, then we can figure that out. I'd be happy to markup your code and offer some possible implementations and feedback if you are interested.

If you disagree with anything I said, then that's ok! We should talk about it and I would like to hear your input.

Otherwise, pending your feedback, I can implement these changes myself, using this comment as a starting point.

Thanks again!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Creeper36 committed:
Added sys.exit codes 6 & 8

Creeper36 committed:
Added sys.exit codes 6 & 8

OwenCochell commented:
Hello,

I appreciate your feedback, and your alternative implementations! No worries about the many messages, as you can see I am also guilty of drawn out responses lol.

I still think both these features (error codes and terse logging) are good, and I will be implementing them in the near future after I work out some options. I also think your implementations you provided here are very helpful, and I will be using them to determine the way forward from here. I'd still like to run some ideas past you just to make sure I am understanding your problem, and that my changes will correctly solve your issues.

# Error Codes

I think we are on the same page for the error codes, but I want to write them out here just to be sure. These codes will allow you to determine if an upgrade was installed, if no new versions are available, and if errors have occurred at any step of the process. Here is my idea for error codes:

- `0` - Script was able to complete (implies no errors, new version found, and update installed if installing is enabled)
- `10` - No new version was found, nothing installed (implies no errors, and you have the latest server version)
- `1` - Network error, was unable to check or download upgrades
- `2` - Filesystem error, was unable to properly manipulate filesystem (can't delete, copy, bad permissions, etc.)
- `3` - Validation error, downloaded upgrade failed integrity check (could fall under network errors)
- `4` - Version error, unable to parse version information from server config
- `5` - Python error, version is too low

If you get a return code of `0`, then a new version was found, and the update was installed correctly. The script will still return `0` if a new version was found and the script is in check mode `-c` (no install preformed). If you have disabled checking via the `-nc` flag, then this script will always jump to the install step, and if it succeeds then code `0` will be returned.

If you get a return code of `10`, then no new version was found, and this means you are running the latest version with the given constraints. Like above, if the script is in check mode and no update was found then it will still return `10`. If checking is disabled, code `10` will never be returned as the script is assuming an update is found. In this case, it will only ever return `0` (install succeeded), or an error code if something goes wrong.

If you get anything else, then an error occurred, and something is wrong (either with this script, your network connection, or filesystem) and needs intervention. Error states can occur under any operation mode.

These codes are kind of broad, and could be split into even finer groups (special error codes for all network/filesystem failure types), but I am unsure if this is worthwhile. I think keeping it simple for now is the way to go. Implementing these codes should be pretty simple and should result in minimal changes. 

The selection of these codes is mostly arbitrary, and if you think a different code makes more sense for a particular situation then I'd like to hear about it. I am choosing these codes from a unix perspective, and the main rule is that if nothing 'goes wrong', then you should return code `0`. I think it's safe to say that if we encounter any fatal errors then we should not return zero, but I also think it's ok to return a non-zero code if an update has not been found. In your implementation, you return code `6` if an update was installed and code `8` if a new version was found. I think in these cases (where things go right), it would be appropriate to return `0` instead. While my proposed solution does not have separate codes for new version found and update installed, I think the result can be still be deduced by taking into consideration the provided flags. So, if you only want to check for a new version and analyze the error code for a result, you can enable checking via the `-c` flag and look for code `0` (new version found) and code `10` (no new version found), and you can go from there.

Does this better align with your needs? I believe this should fit into your use case but I could have missed something.

# Terse Output

I also agree that the filtering strings should be kept in the script, but I do appreciate your alternative implementation. I am still unsure if the filtering approach is the way to go, I think I still prefer explicitly specifying if output is 'terse' enough for each message. How this will be done is not yet clear, I have a few ideas I am looking into. A very simple one would be to alter the `output` function (discussed in my previous message). Another option is the python logging module, which is very flexible and configurable (and will enable cool features like logging to a file without the need for pipes), but this library also adds some complexity and will require a lot of changes. Regardless of how this gets implemented, the results should be nearly identical to your approach: only outputting relevant messages without any fancy graphics or animations. 

I just had this idea, would machine readable output be useful to your workflow? If implemented, it will output information in some readable format that can easily be parsed by any automated tool. For example, the version selection message is currently:

```
Server Version Information:
  > Version: [1.21.1]"
  > Build: [100]"
```

This looks neat (at least in my opinion) and communicates all the necessary info to the user. However, it is somewhat hard for machines to read, so if machine readable output is enabled we would get something like this:

```
{
    'type': 'version_info',
    'out': {
        'version': '1.12.1',
        'build': 100
    }
}
```

This JSON output can easily be parsed and utilized. Any relevant outputs would have data structured like this if machine readable output is enabled. Users can wrap and use this script in whatever setup they like, and parsing the info would be pretty easy. Any thoughts? This PR is getting a bit out of scope, and this feature would require a decent rework on how outputting is done, but it could make automating aspects of this script easier.

# Conclusion

Please let me know your thoughts on these proposals (in particular I want to ensure my error codes are robust enough for your use case). At the end of the day, both these features will be implemented into the project one way or another, but I want to make sure they mesh with your use case. I also want to make sure these are correct at the start, breaking these things down the road can lead to trouble, and could break things. 

Thanks!

Creeper36 commented:
I got No Internet Connection error setup, tested, and working.
TERSE OUTPUT:My whole idea here was to not disrupt the main output.. I am only filtering when --batch is used... It doesnt do anything any other time.. its too simple not to use.. if a script is ran it should run normally.. if a argument is used then that does something.. like turning on the filtering.. I am not trying to add more to your script.. I am trying to keep it as simple as possible.. if you edit a title or something similar to check the filterArray to see if any match or you may want them to match.   Me = Easy ........ You = More Detailed = Better = BUT More Potential For Errors

ERRORLEVEL:0 - Normal Exit- New MC Version Found   
1 - Generic Error- Any Unlisted Error2 - No Internet Connection Found   
3 - Critical Error- Python Version Too Low   
9 - Normal Exit- Server_Update.py was Updated   
10 - Normal Exit- No MC Update   

I have now implemented these errorlevels.



Message ID: ***@***.***>

Creeper36 committed:
Added Internet Connection Available OR Error

Creeper36 commented:
I wanted to try the Python Version below 3.7.0 so I installed 3.6.9.. verified this version.. and it doesnt show any of this >>> if sys.version_info < (3,7,0):

    sys.stdout.write("\n--== [ Invalid python version! ] ==--\n")
    sys.stdout.write("Current version: {}\n".format(version_info))
    sys.stdout.write("Expected version: 3.7+\n")
    sys.stdout.write("\nPlease install the correct version of python before continuing!\n")

    sys.exit(3)
It errors out.. doesnt even get to run the script.. it errors as its loading.. I see no current version or expected version.. none of it.. just wanted you to know.. what the script needs to run it isnt available till 3.7.0... i double checked!
IMessage ID: ***@***.***>

Creeper36 committed:
Previous misdefined exit code

Creeper36 commented:
After reviewing your script, there are no forward references detected, which means from __future__ import annotations is unnecessary. The script does not reference classes or types in type hints before they are defined.

Recommendation:

You can safely remove this line:
from __future__ import annotationsRemoving it will not affect the functionality of the script, even if you're running it on Python 3.7+ or higher. This will also make the script compatible with Python 3.6.
>>>>> I am using python 3.6.8 to test the error detection.. it fails and crashes with "from __future__ import annotations" in place.After I remove this the script does correctly error out with the appropriate comments.
THERE ARE 2 REASONS TO REMOVE THIS!  Consider this?

Creeper36 committed:
Final Touches.. Completed

Creeper36 commented:
Those 4 deprecated commands ( -C , -ndc , -nr , -config ) .. In my opinion.. I would announce on your next release that the following release after that the command will be removed to clean up the script.. again.. just my thoughts.  
Message ID: ***@***.***>

Creeper36 committed:
added more if not args.batch and cleared out the filterarray.. left filterarray at top for future

Creeper36 committed:
Delete server-update.py

Creeper36 committed:
added 'if not args.batch' and clear out filterarray, left at top for future.

Creeper36 commented:
I have cleared out the FilterArray by adding more 'if not args.batch' ... it works very well.. I left the FilterArray = [] at the top in case you want it for future or someone wants to add their own.. i had thought you mentioned that before.I've been having fun.. I do this for fun for my MC server.

PaperMC-Update/server_update.py at master · Creeper36/PaperMC-UpdateMessage ID: ***@***.***>

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
PaperMC-Update/server_update.py at master · Creeper36/PaperMC-Update

Contribute to Creeper36/PaperMC-Update development by creating an account on GitHub.
 |

 |

 |

Creeper36 commented:
ChangeLog:
Added sys.exit(#) .. this number is the Errorlevel #.Moved the contents of FilterArray out into the 'if not args.batch'FilterArray ["word1", "pattern2", etc] will still filter in --batch.. left open at top of script for future use.Added Ping 8.8.8.8 to check if Internet Connection is alive before trying to update. (ERRORLEVEL=2)Removed Progress Bar and Title Image with Names during --batch modeFew minor word changes

| PYTHON server_update.py C:\Minecraft\paper.jar |
|


|    IF %ERRORLEVEL% EQU 10 GOTO SAMEPAPER |
|


|    IF %ERRORLEVEL% EQU 9 GOTO SERVERUPDATE |
|


|    IF %ERRORLEVEL% EQU 3 GOTO ERRORPYTHON |
|


|    IF %ERRORLEVEL% EQU 2 GOTO ERRORINTERNET |
|


|    IF %ERRORLEVEL% EQU 0 GOTO NEWPAPERFOUND |
|

   GOTO SAMEPAPER
















Added --batch and errorlevels information to README.mdMessage ID: ***@***.***>

Creeper36 commented:
hello.. haven't heard from you in a while.. hows things going?  need some help please ask.
ThanksChadhaMessage ID: ***@***.***>

Creeper36 committed:
Huge Update --  [-ua name@email.com] is now requested

The API was updated to restructure how version and build data are retrieved — introducing grouping by major version and reversing the order of results. This broke several assumptions in the original code, which expected a flat, ascending list. As a result, we had to refactor multiple components to correctly parse the new nested structure, handle reversed ordering, and maintain compatibility with the updated versioning system. These changes ensured the updater script continues to function reliably with the latest API format.  All this resulted in the [-ua name@email.com] becoming necessary.

Creeper36 committed:
Update README.md.. --user-agent explained

Creeper36 committed:
Update CHANGELOG.md

Creeper36 committed:
Update README.md

Creeper36 committed:
Update README.md

Creeper36 committed:
Update README.md

Creeper36 committed:
Update README.md

Creeper36 committed:
Major Update ... now argument --user-agent is being requested.

Creeper36 committed:
Update README.md

Creeper36 committed:
Fixed bug where 'latest' build selection could pick build 1 instead of the newest.

- Fixed bug where 'latest' build selection could pick build 1 instead of the newest.
  * Changed --build arg type from str to int for consistent sentinel handling (-1 for latest).
  * Updated _select() to normalize input and select the max() build number rather than relying on list ordering.
- Verified Paper API only returns sequential numeric builds; no beta filtering needed for now.
- Cleaned up duplicate/old _select() implementations and ensured single authoritative version in code.
- Added safeguards:
  * Removed potential cache collision between get() and _get() by isolating raw response caching.
  * Added network timeouts to urllib requests to avoid indefinite hangs.
  * Removed unreachable return statements and redundant code.
- Minor quality-of-life:
  * Batch mode flag (-ba) for cleaner logs.
  * Option for dry-run testing and simulated builds/versions to validate selection without waiting for new releases.

Creeper36 committed:
Merge branch 'master' into master

Creeper36 committed:
removed a stray return

Creeper36 committed:
Added Check for Internet Connection

Added a check for internet connection before trying to search for remote versions.  If not internet connection was found the script would error and crash out.  Going forward, If no internet connection is found the script will not proceed further in the script and exit with a status comment for logging and status code 2.
Moved 'filterArray' to the top for easier edits in the future.

Creeper36 committed:
Added linux support to check internet connection

Creeper36 committed:
Creeper Release 1.0
